File: TypeScript/Multiplayer/PlayerEvent.ts
import { PlayerEventID } from './exports.js';

export type PlayerEvent = PlayerTick
    | PlayerFaster
    | PlayerSlower
    | PlayerPause;

export interface PlayerTick {
    readonly id: PlayerEventID.Tick;
}
export interface PlayerFaster {
    readonly id: PlayerEventID.Faster;
}
export interface PlayerSlower {
    readonly id: PlayerEventID.Slower;
}
export interface PlayerPause {
    readonly id: PlayerEventID.Pause;
}

File: Sources/GameAPI/PlayerEvent.swift
import JavaScriptInterop

enum PlayerEvent {
    case faster
    case slower
    case pause
    case tick
}
extension PlayerEvent: JavaScriptDecodable {
    enum ObjectKey: JSString, Sendable {
        case id
    }

    init(from js: borrowing JavaScriptDecoder<ObjectKey>) throws {
        let id: PlayerEventID = try js[.id].decode()

        switch id {
        case .Faster:   self = .faster
        case .Slower:   self = .slower
        case .Pause:    self = .pause
        case .Tick:     self = .tick
        }
    }
}
